modules/k8s/jsonpath.go

```diff
@@ -0,0 +1,72 @@
+package k8s
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/testing"
+	"github.com/stretchr/testify/require"
+)
+
+// UnmarshalJSONPath allows you to use an arbitrary JSONPath string to query a json blob and unmarshal the resulting
+// output into a go object. Note that the output will always be a list. That means that if you query a single object,
+// the output will be a list of single element, not the element itself. However, if the json path maps to a list, then
+// the output will be that list.
+// Example:
+//
+// jsonBlob := []byte(`{"key": {"data": [1,2,3]}}`)
+// jsonPath := "{.key.data[*]}"
+// var output []int
+// UnmarshalJSONPath(t, jsonBlob, jsonPath, &output)
+// // output is []int{1,2,3}
+//
+// This will fail the test if there is an error.
+func UnmarshalJSONPath(t testing.TestingT, jsonData []byte, jsonpathStr string, output interface{}) {
+	err := UnmarshalJSONPathE(t, jsonData, jsonpathStr, output)
+	require.NoError(t, err)
+}
+
+// UnmarshalJSONPathE allows you to use an arbitrary JSONPath string to query a json blob and unmarshal the resulting
+// output into a go object. Note that the output will always be a list. That means that if you query a single object,
+// the output will be a list of single element, not the element itself. However, if the json path maps to a list, then
+// the output will be that list.
+// Example:
+//
+// jsonBlob := []byte(`{"key": {"data": [1,2,3]}}`)
+// jsonPath := "{.key.data[*]}"
+// var output []int
+// UnmarshalJSONPathE(t, jsonBlob, jsonPath, &output)
+// => output = []int{1,2,3}
+func UnmarshalJSONPathE(t testing.TestingT, jsonData []byte, jsonpathStr string, output interface{}) error {
+	// First, unmarshal the full json object. We use interface{} to avoid the type conversions, as jsonpath will handle
+	// it for us.
+	var blob interface{}
+	if err := json.Unmarshal(jsonData, &blob); err != nil {
+		logger.Logf(t, "Error unmarshaling original json blob")
```

Comment:
Why do we need to log the error we're returning? Seems redundant? Do we need more descriptive error types?

Reply:
I added these in because the error returned by these are not very descriptive, and also because we don't have a stack trace, so it isn't clear where in this routine it failed.

E.g., a json error would show something like "unexpected char {", but coincidentally, the jsonpath parser could emit a very similar error. So you wouldn't exactly know if the error is from json parsing, or jsonpath parsing.

But this isn't ideal either because the error log doesn't show up in the assertion failure so you have scroll through logs. I'll switch to using error types so that the error message shows up in the assert library.

Reply:
Updated ee1205c